### PR TITLE
Split cmp

### PR DESCRIPTION
**Pull request by robdockins**
Propagate consequences of the `split-cmp` branch of in Cryptol.  Follows on to #18

### CODE REVIEW

File: src/Verifier/SAW/Cryptol.hs [hunk 1]
```diff
@@ -99,8 +107,42 @@ bindProp sc prop env = do
...
   v <- scLocalVar sc 0
   k <- scSort sc (mkSort 0)
-  return $ env' { envP = Map.insert (normalizeProp prop) (v, 0) (envP env')
-                , envS = k : envS env' }
+  return $ env' { envP = insertSupers prop [] v (envP env')
+                , envS = k : envS env'
+                }
+
+-- | When we insert a nonerasable prop into the environment, make
+--   sure to also insert all it's superclasses.  We arrange it so
```
Comment (brianhuffman) [inline src/Verifier/SAW/Cryptol.hs]: "its"

File: src/Verifier/SAW/Cryptol.hs [hunk 2]
```diff
@@ -99,8 +107,42 @@ bindProp sc prop env = do
...
+--   that every class dictionay contains the implementation of its
```
Comment (brianhuffman) [inline src/Verifier/SAW/Cryptol.hs]: "dictionary"

File: src/Verifier/SAW/Cryptol.hs [hunk 3]
```diff
@@ -99,8 +107,42 @@ bindProp sc prop env = do
...
+      C.PRing      -> super C.PZero "ringZero" t m
+      C.PLogic     -> super C.PZero "logicZero" t m
+      C.PField     -> super C.PRing "fieldRing" t m
+      C.PIntegral  -> super C.PRing "integralRing" t m
+      C.PRound     -> super C.PField "roundField" t . super C.PCmp "roundCmp" t $ m
+      C.PCmp       -> super C.PEq "cmpEq" t m
+      C.PSignedCmp -> super C.PEq "signedCmpEq" t m
+      _ -> m
+ go _ = m
+
```
Comment (brianhuffman) [inline src/Verifier/SAW/Cryptol.hs:145]: This looks good for the current set of superclass relationships. But I wonder how well the same approach would work if we ever get any diamonds in our superclass graph in the future? I don't know if this is something we need to worry about.

File: src/Verifier/SAW/Cryptol.hs [hunk 4]
```diff
@@ -378,16 +431,48 @@ proveProp sc env prop =
...
+        -- instance (fin n, Eq a) => Eq [n]a
+        (C.pIsEq -> Just (C.tIsSeq -> Just (n, a)))
+          -> do n' <- importType sc env n
+                a' <- importType sc env a
+                pa <- proveProp sc env (C.pEq a)
+                scGlobalApply sc "Cryptol.PEqSeq" [n', a', pa]
+        -- instance Eq ()
+        (C.pIsEq -> Just (C.tIsTuple -> Just []))
+          -> do scGlobalApply sc "Cryptol.PEqUnit" []
+        -- instance (Cmp a, Cmp b) => Cmp (a, b)
```
Comment (brianhuffman) [inline src/Verifier/SAW/Cryptol.hs]: `(Eq a, Eq b) => Eq (a, b)`

File: src/Verifier/SAW/Cryptol.hs [hunk 5]
```diff
@@ -56,7 +58,10 @@ import Verifier.SAW.TypedAST (mkSort, mkModuleName)
 data Env = Env
   { envT :: Map Int    (Term, Int) -- ^ Type variables are referenced by unique id
   , envE :: Map C.Name (Term, Int) -- ^ Term variables are referenced by name
-  , envP :: Map C.Prop (Term, Int) -- ^ Bound propositions are referenced implicitly by their types
+  , envP :: Map C.Prop (Term, [FieldName], Int)
+              -- ^ Bound propositions are referenced implicitly by their types
+              --   The actual class dictionary we need is obtained by applying the
+              --   given field selctors (in reverse order!) to the term.
```
Comment (brianhuffman) [inline src/Verifier/SAW/Cryptol.hs]: "selectors"

File: src/Verifier/SAW/Cryptol.hs (hunk 3 above)
Comment (robdockins) [reply src/Verifier/SAW/Cryptol.hs:145]: If there's diamonds in the graph, I guess you'll get a different field access path depending on exactly what order you explore the graph in.  If all our instances are confluent, it shouldn't matter, semantically.  I guess we could always prefer the access path with fewest fields, but you'd still have arbitrary choices to make for same-length paths.

**Pull request merged**